src/file_handler.rs

```diff
@@ -0,0 +1,328 @@
+use crate::{disk_backed_queue::DiskBackedQueue, interner::Interner, ortho::Ortho, FoldError};
+use std::fs;
+use std::path::Path;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+const HEARTBEAT_GRACE_PERIOD_SECS: u64 = 600; // 10 minutes
+
+pub fn recover_abandoned_files(input_dir: &str, in_process_dir: &str) -> Result<(), FoldError> {
+    let in_process_path = std::path::Path::new(in_process_dir);
+    
+    if !in_process_path.exists() {
+        return Ok(());
+    }
+    
+    let mut recovered_count = 0;
+    
+    // Check for folders with stale heartbeats in in_process
+    for entry in fs::read_dir(in_process_path).map_err(|e| FoldError::Io(e))? {
+        let entry = entry.map_err(|e| FoldError::Io(e))?;
+        let entry_path = entry.path();
+        
+        if entry_path.is_dir() {
+            let heartbeat_path = entry_path.join("heartbeat");
+            
+            // Check if this folder has a heartbeat file
+            if heartbeat_path.exists() && is_heartbeat_stale(&heartbeat_path)? {
+                // Check if this is a txt.work folder
+                let source_txt_path = entry_path.join("source.txt");
+                if source_txt_path.exists() {
+                    // Recover the txt.work folder back to input as a plain txt file
+                    if let Some(folder_name) = entry_path.file_name() {
+                        let folder_name_str = folder_name.to_str().unwrap_or("recovered");
+                        if folder_name_str.ends_with(".txt.work") {
+                            let base_name = &folder_name_str[..folder_name_str.len() - 9]; // Remove ".txt.work"
+                            let target_path = format!("{}/{}.txt", input_dir, base_name);
+                            fs::rename(&source_txt_path, &target_path).map_err(|e| FoldError::Io(e))?;
+                            println!("[fold] Recovered abandoned txt file: {} -> {}", folder_name_str, target_path);
+                            
+                            // Remove the abandoned work folder
+                            fs::remove_dir_all(&entry_path).map_err(|e| FoldError::Io(e))?;
+                            recovered_count += 1;
+                        }
+                    }
+                }
+                // If it's an archive folder with stale heartbeat, just remove it
```

Comment:
This may be pointing to an issue - make sure that, in the same way that in process txt filters are moved for mutual exclusion, in process archives are moved too. Make sure those are properly restored when stale.